core/domain/opportunity_services.py

```diff
@@ -444,6 +444,42 @@ def get_skill_opportunity_from_model(model):
         model.id, model.skill_description, model.question_count)
 
 
+def get_skill_opportunities_with_topic(cursor):
+    """Returns a list of skill opportunities available for questions with topic
+    information.
+
+    Args:
+        cursor: str or None. If provided, the list of returned entities
+            starts from this datastore cursor. Otherwise, the returned
+            entities start from the beginning of the full list of entities.
+
+    Returns:
+        3-tuple(opportunities, cursor, more). where:
+            opportunities: list(dict). A list of dict of opportunity details.
+            cursor: str or None. A query cursor pointing to the next
+                batch of results. If there are no more results, this might
+                be None.
+            more: bool. If True, there are (probably) more results after
+                this batch. If False, there are no further results after
+                this batch.
+    """
+    topics_with_skills = topic_fetchers.get_all_topics_with_skills()
+    skill_opportunities, cursor, more = get_skill_opportunities(cursor)
+    id_to_skill_opportunity = _skill_id_to_skill_opportunity_dict(
+        skill_opportunities)
+    opportunities = []
+    for topic in topics_with_skills:
+        for skill_id in topic.get_all_skill_ids():
+            if len(opportunities) == feconf.OPPORTUNITIES_PAGE_SIZE:
+                break
+            if skill_id in id_to_skill_opportunity:
+                skill_opportunity = id_to_skill_opportunity[skill_id]
+                skill_opportunity_with_topic = skill_opportunity.copy()
+                skill_opportunity_with_topic['topic_name'] = topic.name
```

Comment:
Maybe handel populating topic name in the controller layer. In general, the service should only return objects!

Reply:
Do you mind clarifying what you mean here? This method is adding the topic name to the dict representation of the skill opportunity so that the topic name is associated with the opportunity for use in the frontend. We would still need to pass along the topic information mapped to each opportunity some way.

Reply:
@sagangwee Basically, the service layer should only return domain objects (instead of dict) and the other layers can use it and modify it to a dict if required!

In our case, the `opportunity_services.py` should return a domain object and the controller which is using this service will create a dict out of it and populate the dict as per requirements.

Reply:
Hmm, but don't we return dict objects in `opportunity_services.py` already? e.g. `get_translation_opportunities` or `get_exploration_opportunity_summaries_by_ids`. I'm wondering what the distinction is here.

Reply:
Hm, if that's the case, those should be changed too. We should only be passing domain objects around, not raw dicts.